### ISSUE TITLE
[Policy] Commitment on new Minimal development test

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest#9645]
This issue is for discussing a new policy regarding the Minimal development test.

It is closely tied to the complete Minimal overhaul I submitted in this PR: #9450
This issue kinda sorta assumes the Minimal overhaul is merged.
But we can already discuss policy right now, even before the merge happens.

So. Here's a new policy I propose:

##### New policy for Minimal development test

1. When adding a Lua API feature in the engine, add a test in Minimal (if that's practical to do)
2. When a bug in the engine is fixed, consider adding a regression test in Minimal (if that's practical to do)
3. For Minimal, follow the conventions laid out in the `README.md` in Minimal
4. The quality standards for merging PRs that concern Minimal *only* are lower:
    * Only 1 core dev approval
    * Code stlye and other formal criteria do not matter *that* much
    * Code doesn't even need to work perfectly, but half-finished code needs to be put into `experimental`

[END OF PROPOSED POLICY]

##### Rationale

1. Minimal has been neglected for years, which resulted in Minimal becoming horribly outdated and of very limited use. By making it a requirement to add test cases, when appropriate, we guarantee that Minimal stays up-to-date (well, at least we TRY). With an up-to-date Minimal it should also be faster to reveal bugs and—even more important—regressions. Having a test requirement might also be helpful to catch bugs or even design flaws in new features BEFORE they are merged. In the long term, my hope is that the code quality of Minetest overall improves because it faciliates a culture of testing. This hope is based on experience. When developing the Minimal overhaul in #9450 I have uncovered (and reported) numerous bugs in the engine that were new to me. I therefore expect that such a policy will help us catch even more bugs over time.
2. Same as 1.
3. (explained in #9450)
4. Basically, Minimal is the developer's playground, so it's OK if it gets a *little* “dirty” in there. Having *some* kind of test cases for some feature is still much better for the engine quality in the long term than having none at all. Being perfectionistic about Minimal can be bad to Minetest longterm, because important and useful test cases are being delayed too long, in which time bugs can creep in more easily, and it is harder for people to test out stuff. My hope is that PRs for Minimal will get merged much faster, because there are fewer reasons to block a PR from getting merged. That still doesn't mean that we shouldn't try for good code in Minimal, but it's no longer a strict requirement.

##### Important note

Do not adopt this policy without thinking or if there just isn't an agreement! Adopting this policy is a commitment for everyone involved in Minetest, and it won't make sense if it's adopted only on paper but has no real-world impact.

I'm not a core dev myself, so you will probably disagree about some or even all of the points I just listed. The point is to discuss this stuff.

**Comment by Wuzzy2** [comment]
Bump. As Devtest has been merged, this dission is now relevant.

**Comment by paramat** [comment]
> The quality standards for merging PRs that concern Minimal only are lower:
Only 1 core dev approval
Code stlye and other formal criteria do not matter that much
Code doesn't even need to work perfectly, but half-finished code needs to be put into experimental

I disapprove all these.

1 core dev approval is similar to our 'trivial rule', where a core dev can approve and merge alone but only if announced on IRC 15 mins before (so that other core devs can check what was done).
Code in devtest is not always trivial, usually it will not be, so we cannot say 1 approval is always enough.
I think the existing 'trivial' rule is sufficient here. We always use our judgement on triviality and devtest will have more trivial stuff, so we will naturally be less strict.
So, i think no new rules or changed rules are needed.

Devtest code readability is just as important as any other MTG or engine code.
"do not matter that much" is not possible, codestyle is either enforced or are not.

Broken (the real meaning of 'not working perfectly') or half-finished code is usually unacceptable.
If code is half-finished it should be finished before merging, there is no point merging it any earlier. Merging half-finished code that requires a follow-up PR just creates mess and increases everyone's workload.
Again, we can use our judgement, as we already do, and judge this per situation, no new rules are needed.

**Comment by Wuzzy2** [comment]
I can accept this compromise.

**Comment by Wuzzy2** [comment]
@paramat, can you please invite other devs to the discussion?

**Comment by sfan5** [comment]
I pretty much already follow point 1 and 2, can't comment on point 3 as I haven't read devtests' README.

I disagree with the lower standards suggested in point 4 too.

**Comment by Wuzzy2** [comment]
Another reason for point 4 is to speed up merging of devtest changes. It's not good development if devtest PRs are rotting too long. But on the other hand there hasn't been a demand for devtest changes lately anyway. Not even from me. So maybe this policy change isn't needed after all. I think devtest, for now, is fine.

Frankly, I don't think points 1 and 2 have been seriously followed in the past. Sure, *some* features immediately got test cases, but a ton were missing. Remember, I had to completely overhaul Devtest to get it up-to-date again, a lot of features had no test cases whatsoever.
What I was getting here at is that it should be a *condition for inclusion* for a given feature. No new feature without some kind of test case in Devtest. As long a test case makes sense, of course. Similar to the documentation requirement. No new API stuff canget accepted without updating the docs.

For example, recently the `allfaces` drawtype got `visual_scale` support, way after Devtest got merged. Devtest was NOT updated in the corresponding PR. I had to add test nodes myself, after-the-fact … #10225

This is what I'm talking about. If Minetest does not make adding Devtest test cases a merge condition, then Devtest will be doomed to become outdated very very fast. Neglecting the old Minimal was a big reason why it was in such a bad shape.
I'm not saying test cases in Devtest need to be added for absolutely everything. There are cases in which there can't be any reasonable change in Devtest. That's OK. But “I could not be bothered” should no longer be a valid excuse. :P


BTW: Here's the Devtest README: https://github.com/minetest/minetest/blob/master/games/devtest/README.md (only the final section, "Design Philosophy", is relevant)

**Comment by rubenwardy** [comment]
I've been asking that any formspec PRs update the test formspecs, as it's very useful

**Comment by paramat** [comment]
> What I was getting here at is that it should be a condition for inclusion for a given feature. No new feature without some kind of test case in Devtest. As long a test case makes sense, of course.

I disagree. It should be strongly encouraged and asked for of course, and i expect many developers will feel obliged to do this anyway.
But it would be very wrong to block perfectly good functional work because adding test code to devtest (which unlike documentation is not essential) has not been done. There may be valid reasons why someone cannot include devtest test code, maybe they would like to but have no time for that, maybe a devtest test is beyond their ability or they do not know how to approach that.
I understand why you want to make it compulsory, to make sure it is done, but this will inevitably result in stupid and unnecessary blocking of good work, as well as infuriating the contributors involved.

**Comment by rubenwardy** [comment]
however, there's plenty of time of code style changes, of course

**Comment by SmallJoker** [comment]
Proposal, covering most of the current practices:

 * Add a test to Minimal in bugfix/feature PRs when it makes sense and helps to avoid regressions in the future
 * Minimal PRs still require 2 approvals (unless marked as "trivial")
 * Because MInimal does not affect regular gameplay, Minimal-only PRs may be marked as "trivial" easier
 * Lower code style requirements

This should be collected for a `contributions.md` file (linked in the dev wiki).

Discussion: http://irc.minetest.net/minetest-dev/2020-08-01#i_5718865

**Comment by Wuzzy2** [comment]
I'm happy with this.

**Comment by Wuzzy2** [comment]
I'm happy with this.

**Comment by rubenwardy** [comment]
> This should be collected for a `contributions.md` file (linked in the dev wiki).

The file is here: https://github.com/minetest/minetest/blob/master/.github/CONTRIBUTING.md

**Comment by Wuzzy2** [comment]
We have a CONTRIBUTING file? I didn't know! Well, no wonder that I didn't see it, it's hidden inside the hidden `.github` directory. -_-

If you just browse the repo outside of GitHub, you normally wouldn't find it. Most people would not bother looking inside hidden (well, no really, but you know what I mean!) dirs … Why does GitHub have to be so weird about this? XD

**Comment by rubenwardy** [comment]
There's a link to that file when you open your first PR
